### PR TITLE
Multisig fix

### PR DESCRIPTION
**Pull request by onvej-sl**
* https://github.com/trezor/trezor-firmware/pull/4351/commits/d99fea38d791a44b4add9f2c7a0e67a84fd36318 and https://github.com/trezor/trezor-firmware/pull/4351/commits/cae49da30e5dca0c5e006088aa282a85f9f89e77 disallow per-node paths which fixes https://github.com/satoshilabs/trezor-firmware/issues/191.
* https://github.com/trezor/trezor-firmware/pull/4351/commits/d19cbfa3cbfa1252c75184ae5eb1f298f519305d introduces sorting of pubkeys in multisig which fixes https://github.com/satoshilabs/trezor-firmware/issues/193.
* https://github.com/trezor/trezor-firmware/pull/4351/commits/267c1c7aaac3fcefed0aeac994d7f58f65d23502 removes a path that is not used anymore (see https://satoshilabs.slack.com/archives/C01L1HT4B46/p1730908353546419). This was the only allowed path that doesn't end with `/change/address_index`. This caused some paths weren't properly recognized as change outputs. However, the main reason for removing this path is primarily to prevent further problems in the future.
* https://github.com/trezor/trezor-firmware/pull/4351/commits/d9b7a57e8b478fa8fa3be3463f08f242d95c4936 forbid multisig to singlesig change outputs. This allows users to have greater control over where their funds are sent because since https://github.com/trezor/trezor-firmware/pull/2925, we also display the path for internal outputs.  Nevertheless, this doesn't always work as expected, see https://github.com/trezor/trezor-firmware/issues/4319).

None of these changes were backported to legacy.

### CODE REVIEW

**Comment by onvej-sl** [comment]
Related issue: https://github.com/trezor/trezor-firmware/issues/4356

**Comment by onvej-sl** [comment]
@andrewkozlik please do a high-level review, ensuring particularly that this pull request fixes the mentioned issue.
@mmilata please do a generic review.

File: core/src/apps/bitcoin/sign_tx/matchcheck.py
Comment (mmilata) [inline core/src/apps/bitcoin/sign_tx/matchcheck.py:114]: Left hand side is always true, is it intentional?

File: core/src/trezor/ui/layouts/mercury/__init__.py
Comment (mmilata) [inline core/src/trezor/ui/layouts/mercury/__init__.py:146]: Nit: maybe capitalize as XPUBs?

**Review by mmilata — COMMENTED** [review]
Looking good, found some nits, another pair of eyes on `apps/bitcoin/signing/` would be nice.

File: core/failing_test
Comment (onvej-sl) [reply core/failing_test]: Removed in https://github.com/trezor/trezor-firmware/pull/4351/commits/b9ca79b2c1c1559d42fa0e5cb6e4e7fcc696214f.

File: core/src/apps/bitcoin/sign_tx/matchcheck.py
Comment (onvej-sl) [reply core/src/apps/bitcoin/sign_tx/matchcheck.py]: Not intentional, fixed in https://github.com/trezor/trezor-firmware/pull/4351/commits/7ed1c424dfdd7a4619c811763a28e2d78f0743dd.

File: legacy/firmware/transaction.c
Comment (onvej-sl) [reply legacy/firmware/transaction.c:404]: Definitely. Fixed in https://github.com/trezor/trezor-firmware/pull/4351/commits/e7277e259129ff0569e84ad2ddfa27d6bac9072a.

File: legacy/firmware/transaction.c
Comment (onvej-sl) [reply legacy/firmware/transaction.c:455]: Fixed in https://github.com/trezor/trezor-firmware/pull/4351/commits/e7277e259129ff0569e84ad2ddfa27d6bac9072a.

File: core/.changelog.d/4351.added
Comment (onvej-sl) [reply core/.changelog.d/4351.added:1]: We can also mention [BIP383](https://github.com/bitcoin/bips/blob/master/bip-0383.mediawiki) or [BIP67](https://github.com/bitcoin/bips/blob/master/bip-0067.mediawiki). I'm hesitant to mention descriptors because I think we [do not support](https://github.com/trezor/trezor-firmware/blob/onvej-sl/multisig-fix/core/src/apps/bitcoin/get_public_key.py#L123) multisig descriptors at all. Anyway, "sortedmulti" itself might be somewhat misleading as well.

What about "Added support for lexicographic sorting of pubkeys"?

Comment (mmilata) [reply core/.changelog.d/4351.added:1]: Sounds good, maybe even _... multisig pubkeys_ or _... pubkeys for multisig_?

File: core/.changelog.d/4351.changed.1
Comment (andrewkozlik) [inline core/.changelog.d/4351.changed.1:1]: I think the message should read "Forbid ..." or possibly "Forbade ...".

File: core/src/apps/bitcoin/multisig.py
Comment (andrewkozlik) [inline core/src/apps/bitcoin/multisig.py:33]: Add comment explaining why we sort the XPUBs.

File: core/src/apps/bitcoin/common.py
Comment (andrewkozlik) [reply core/src/apps/bitcoin/common.py:270]: Add comment explaining the two scenarios that may be going on here. Also we may want to get rid of `if`.

File: core/src/apps/bitcoin/get_address.py
Comment (andrewkozlik) [inline core/src/apps/bitcoin/get_address.py:117]: typo "addresss"

Comment (andrewkozlik) [reply core/src/apps/bitcoin/get_address.py:117]: Also, how about adding a bit more to the rationale, along the lines that malicious software on the host computer could use this to execute a ransom attack.

File: core/.changelog.d/4351.changed.3
Comment (andrewkozlik) [inline core/.changelog.d/4351.changed.3:1]: Also "Forbid ..." or "Forbade ...", preferably the former. Also this concerns only change outputs and GetAddress, so I'd be specific about that. Maybe "Forbid different sub-paths in multisig change-outputs and receive addresses."

File: core/src/apps/bitcoin/sign_tx/matchcheck.py
Comment (andrewkozlik) [inline core/src/apps/bitcoin/sign_tx/matchcheck.py:107]: Since we removed the logic which accepts singlesig change outputs even when multisig inputs are present, we can simplify the code a little bit by doing the following here:
```py
if not txio.multisig:
    return bytes()
```
What this would do is allow us to:
1. get rid of the special multisig handling in `ChangeDetector.def output_is_change()`, where we only check the fingerprint for multisig outputs, and
2. get rid of the `MultisigChecker` altogether.

File: core/src/apps/bitcoin/sign_tx/matchcheck.py
Comment (andrewkozlik) [inline core/src/apps/bitcoin/sign_tx/matchcheck.py:114]: I think it's perfectly clean to hash the `pubkeys_order` attribute into the multisig fingerprint, since the fingerprint already contains several attributes of `MultisigRedeemScriptType` (XPUBs and threshold). Then we can get rid of `PubkeysOrderChecker` while maintaining the same behavior.

File: core/src/apps/bitcoin/common.py
Comment (onvej-sl) [reply core/src/apps/bitcoin/common.py]: Done in https://github.com/trezor/trezor-firmware/pull/4351/commits/edd0fcab3d910c606952fee325a74fa3cfff424a.

File: core/src/apps/bitcoin/multisig.py
Comment (onvej-sl) [reply core/src/apps/bitcoin/multisig.py]: Done in https://github.com/trezor/trezor-firmware/pull/4351/commits/edd0fcab3d910c606952fee325a74fa3cfff424a and https://github.com/trezor/trezor-firmware/pull/4351/commits/c703cacdcce782dacc8b41c4302ea95ee0f0760a.

Comment (onvej-sl) [reply core/src/apps/bitcoin/multisig.py]: Done in https://github.com/trezor/trezor-firmware/pull/4351/commits/4e52009e889fca6cc7f8f2905c9b3ea2250ea429.

File: core/src/apps/bitcoin/sign_tx/matchcheck.py
Comment (onvej-sl) [reply core/src/apps/bitcoin/sign_tx/matchcheck.py]: Done in https://github.com/trezor/trezor-firmware/pull/4351/commits/4e52009e889fca6cc7f8f2905c9b3ea2250ea429.

File: core/src/apps/bitcoin/sign_tx/matchcheck.py
Comment (mmilata) [reply core/src/apps/bitcoin/sign_tx/matchcheck.py:107]: I'm generally fan of explicit code but this seems fine, especially if you add a comment.

File: core/.changelog.d/4351.changed.1
Comment (onvej-sl) [reply core/.changelog.d/4351.changed.1]: Done in https://github.com/trezor/trezor-firmware/pull/4351/commits/54e4e27910b05c33658dfe5dd2f7d4ed7448c97d.

File: core/.changelog.d/4351.changed.3
Comment (onvej-sl) [reply core/.changelog.d/4351.changed.3]: Done in https://github.com/trezor/trezor-firmware/pull/4351/commits/c5200261fbe0fa28d0c2f826f472e1336cba674c.

File: core/src/apps/bitcoin/sign_tx/matchcheck.py
Comment (onvej-sl) [reply core/src/apps/bitcoin/sign_tx/matchcheck.py:111]: Done in https://github.com/trezor/trezor-firmware/pull/4351/commits/b7d26c794b8839fb9a35208d3b7a4ae336424480.

**Review by andrewkozlik — APPROVED** [review]
High level this LGTM. Good job! Awaiting final approval from @mmilata.

@Hannsek we will need 3rd party multisig providers (Unchained, Casa, ...) to test that these changes do not break their setups. I believe you have the full set of contacts. Please coordinate with them before the next firmware release.

**Review by mmilata — APPROVED** [review]

**Comment by andrewtoth** [comment]
Doesn't this break [`multisig_pubkey_index` ](https://github.com/trezor/trezor-firmware/pull/4351/files#diff-c528ee9407682668ced569943303086cb4098504c4a3684aa85b2f766b283763R60) if using lexigraphic order, since that will return the index of the pubkey if unsorted?

**Comment by andrewtoth** [comment]
That's because that test does not use that function in a breaking way. That is testing unsigned txs. The `multisig_pubkey_index` function breaks when it is used for inserting the signatures.

**Comment by andrewtoth** [comment]
I have it fixed in https://github.com/trezor/trezor-firmware/pull/4159/commits/d06623218f75e87a1f27a904cf3d476e5be57a37

**Comment by onvej-sl** [comment]
> That's because that test does not use that function in a breaking way. That is testing unsigned txs. The multisig_pubkey_index function breaks when it is used for inserting the signatures.

I'm not sure if this is exactly what you're saying, but I discovered that the test fails to catch it because I use empty signatures in `MultisigRedeemScriptType`.

> I have it fixed in https://github.com/trezor/trezor-firmware/commit/d06623218f75e87a1f27a904cf3d476e5be57a37

The fix is not completely correct because `multisig_pubkey_index` is used [here](https://github.com/trezor/trezor-firmware/blob/748a19ac5d6e81c8ab22b00fa1ae467712f07a35/core/src/apps/bitcoin/get_address.py#L113) to detect which xpub belongs to you and the xpubs are not sorted.

**Pull request merged**
